src/NUnitTestExtractor/NUnitTestExtractorApp.cs

```diff
@@ -137,314 +189,108 @@ private static bool IsValidTestMethod(MethodInfo method)
-        /// <param name="categories">categories to add to list</param>
-        private static void AddMethodCategoriesToList(List<string> categoryNames, IEnumerable<Attribute> categories)
-        {
-            foreach (var category in categories)
+            foreach (MethodInfo test in testSuite.GetMethods())
             {
-                CategoryAttribute categoryAttr = (CategoryAttribute)category;
+                var attributes = test.GetCustomAttributesData();
 
-                categoryNames.Add(categoryAttr.Name);
-            }
-        }
+                // if the function attributes don't have either"Test" or "TestCase" attribute, it is not a valid NUnit test, skip it.
+                if (!IsValidTestMethod(test))
+                {
+                    continue;
+                }
 
-        /// <summary>
-        /// Used to determine whether or not the current test/testcase should be included 
-        /// </summary>
-        /// <param name="categoryNames">List of categories to compare info to</param>
-        /// <param name="includeInfo">string of info used to compare to categoryNames</param>
-        /// <returns>returns false if includeInfo is inside of the categoryNames List, true otherwise</returns>
-        public static bool Include(IReadOnlyList<string> categoryNames, string includeInfo)
-        {
-            if (!string.IsNullOrEmpty(includeInfo))
-            {
-                //Remove quotes from string in order to properly process it
-                includeInfo = includeInfo.Replace("\"", "");
+                // search the category attributes for the function
+                var categories = attributes.Where(x => x.AttributeType.Equals(typeof(CategoryAttribute)));
 
-                List<string> includeInfoValues = new List<string>(includeInfo.Split(','));
-                
-                foreach (string value in includeInfoValues)
+                // match the "Exclude" category from the function category attribute list
+                // if the function has matched exclude category, we will skip this function
+                if (categories.Any(x => x.ConstructorArguments.Any(y => y.Value.Equals(_options.ExcludeInfo))))
                 {
-                    if (categoryNames != null && !categoryNames.Contains(value))
-                    {
-                        return false;
-                    }
+                    continue;
+                }
+
+                // match the "Include" category from the function category attribute list
+                // if the function has matched include category, we will count this function as a valid test
+                if (string.IsNullOrWhiteSpace(_options.IncludeInfo)
+                    || categories.Any(x => x.ConstructorArguments.Any(y => y.Value.Equals(_options.IncludeInfo))))
+                {
+                    tests.Add(test.DeclaringType.GetTypeInfo().FullName + "." + test.Name);
```

Comment:
This is only checking the include/exclude categories on the test-level, but it doesn't check the categories on the class-level.